Add WS RPC, Messages and RemoteEnclave

ggwpez opened the pull request:
Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:art: [tee] Move Transaction type and prepare for wire.

Move Transaction from tee.go to transaction.go.
Add Sig type and add custom json en/decoding for
Balance so that big.Ints are encoded as string.
Prepare tee types for the wire by adding json tags.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:white_check_mark: [tee/test] Add random generators and testing.

Add random generators for testing and test the json en/decoding.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [client] Add and use RPC client.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [op] Add and use WireAPI, RPCServer and RPCOperator.

Add the RPCServer struct which serves WS connections and parses
messages. It then forwards the calls to the RPCOperator who
handles proof subscriptions and adds transactions.
The Operator starts and maintains the RPCOperator, such that
the client is always served the latest proofs.

The WireAPI interface is needed for mocking only.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:white_check_mark: [op/test] Mock Enclave and WireAPI.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:white_check_mark: [op+test] Update tests to new enclave/operator semantic.

BlackBox operator_test.go to prevent import cycles.
Use new client.RPC in user.go.
Export EthClient in Operator to enable BB testing.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:white_check_mark: [client] Add mocked RPC integration test.

Test the RPC connection by using a mocked Enclave and WireAPI.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:wrench: [op,client] Prepare config files and logic.

Use 8401 as default RPC port.
Add OpHost and OpPort options to the client.
Add RPCHost and RPCPort options to the operator.
Expose GanacheCommand() to allow for BB testing.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [wire] Add wire package.

Add wire messages and json en/decode test.
Add GenericJSONMarshallingTest.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

sebastianst commented on wire/message.go:
```diff
@@ -0,0 +1,92 @@
+		Tx tee.Transaction `json:"tx"`
+	}
+
+	// Subscribe subscribes to a topic.
```
"Subsribe sets up a client subscription"
we don't subscribe to individual topics.

sebastianst commented on tee/test/rng.go:
```diff
@@ -0,0 +1,60 @@
+
+// RandomSig produces invalid ECDSA signatures, but does not need a wallet
+// to work.
+func RandomSig(rng *rand.Rand) tee.Sig {
```
only need to pass `io.Reader`.

sebastianst commented on tee/test/rng.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/perun-network/erdstall/tee"
+)
+
+func RandomTx(require *require.Assertions, rng *rand.Rand) *tee.Transaction {
```
why don't you, as usual, pass a `t require.TestingT`? Looks strange.

ggwpez replied:
It only needed the require, but i can change that.

sebastianst commented on tee/transaction_test.go:
```diff
@@ -32,3 +35,22 @@ func TestTransaction_SignVerify(t *testing.T) {
+	sender := ttest.NewRandomAddress(rng)
+	receiver := ttest.NewRandomAddress(rng)
+
+	tx := tee.Transaction{
```
why don't you just use `RandomTx` here but manually assemble a random tx?

ggwpez replied:
Good point

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
+	}
+	conn.SetCloseHandler(func(code int, err string) error {
+		rpc.Log().WithField("code", code).Errorf("WS connection was closed: %s", err)
+		rpc.Close()
+		return nil
```
do you on purpose not `return rpc.Close()`?

ggwpez replied:
Something with the linters must be wrong, they should detect this... i changed it

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
+		if result.Error != "" {
+			errChan <- fmt.Errorf("SendTx RPC result: %s", result.Error)
+		} else {
+			close(errChan)
```
I think it would be more consistent to also just send a `nil` here instead of closing the channel.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
+
+// GetDepositProof blocks until it can return the next deposit proof from the
+// operator or an error if the context ran out.
+func (r *RPC) GetDepositProof(ctx context.Context) (tee.DepositProof, error) {
```
Getters usually don't have `Get` in their name in Go. I'd either remove the `Get` or use `Next` as a prefix, then acting more like an iterator. Same for BalanceProofs.

ggwpez replied:
Oh right, this was a remnant from the the old code.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
+		if result.Error != "" {
+			errChan <- fmt.Errorf("Subscribe RPC result: %s", result.Error)
+		} else {
+			close(errChan)
```
again, I'd just push a `nil` here for consistency.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
+// Subscribe subscribes to the Balance and Deposit proof topic.
+// The user must always read the proofs from via `GetDepositProof` and
+// `GetBalanceProof`.
+func (r *RPC) Subscribe(ctx context.Context, user common.Address) error {
```
We could also change the API to let `Subscribe` return an object that acts like an iterator for balance and deposit proofs, having the `BalanceProof` and `DepositProof` methods. Then we have less method cluttering on the `RPC`, and possibly also less state in it? Because we can then move the channels into it. Better encapsulation.

ggwpez replied:
I will try it out  
PS: Done

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
-	return reply, r.call(ctx, "RemoteEnclave.GetBalanceProof", user, &reply)
+func (r *RPC) callCallback(result wire.Result, data []byte) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
```
You should unlock the mutex already after retrieving the callback.Otherwise the mutex is locked until the callback got executed.

sebastianst replied:
You can also change the mutex to a `RWMutex` and only read here. And I realize that you don't delete callbacks, so this map grows indefinitely. So after calling `cb`, you should write-lock the map again and delete the callback.

ggwpez replied:
Right, i added the delete logic

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
-				return err
-			}
-			return ctx.Err()
+func (r *RPC) handleTopic(topic wire.Topic, data []byte) {
```
If you create a subscription class, you can move this method to it.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,229 @@ package client
 	}
 }
+
+func (r *RPC) nextID() wire.ID {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.id++
+	return wire.ID(strconv.Itoa(r.id))
+}
```
seeing this, I'm not sure if it makes sense to have the common mutex `mtx` for ids and callbacks. Why not use an atomic `uint64` here instead?

ggwpez replied:
Did not think about it :laughing:   
PS: Done

sebastianst commented on operator/rpc.go:
```diff
@@ -1,46 +1,182 @@
-}
+type (
+	// RPC handels RPC request and forwards them to the enclave.
+	RPC struct {
```
Maybe rename to `RPCServer`, it's unclear what it is

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+
+type (
+	RemoteOperator interface {
+		AddTransaction(tee.Transaction) error
```
Rename to `Send`

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+	}
+}
+
+func (o *RPCOperator) AddTransaction(tx tee.Transaction) error {
```
Rename to `Send`

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+)
+
+type (
+	RemoteOperator interface {
```
RPC(Server), RPCOperator, RemoteOperator all gets a bit confusing...
Maybe we call the `RemoteOperator` `WireOperator` or `WireAPI` or `WireInterface` instead?

ggwpez replied:
Yes i was having trouble to find good names, will try em out.  
PS: Used `WireAPI` now

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+		SubscribeProofs(common.Address) (ProofSub, error)
+	}
+
+	// RPCOperator is the will be exposed to the client via websocket RPC.
```
```suggestion
	// RPCOperator will be exposed to the client as a websocket RPC Server.
```

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	// ProofSub is a subscription on TEE proofs.
+	ProofSub struct {
+		deposits <-chan tee.DepositProof
+		balances <-chan tee.BalanceProof
+		quit     chan struct{}
+	}
+
+	subAndInputs struct {
+		sub  *ProofSub
+		deps chan tee.DepositProof
+		bals chan tee.BalanceProof
+	}
```
From what I can see, you don't need this unnecessary wrapper type `subAndInputs`, just use a `map[common.Address]ProofSub` in the `RPCOperator`. It's enough to expose the channels as read-only channels in the getters `ProofSub.Deposits` and `Balances`. The class fields can just be normal channels in `ProofSub`.

ggwpez replied:
Ok thanks for the heads up, i investigated the issue again and could remove it.  
I already tried it before but failed, the solution was to expose the `quit` channel by a `Closed()` getter.

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// NewProofSub returns a new proofSub with the given input channels.
+func NewProofSub(deposits <-chan tee.DepositProof, balances <-chan tee.BalanceProof) *ProofSub {
```
I think we don't need to export this constructor. The user of the `operator` package should never create subs, right?

sebastianst commented on operator/rpc.go:
```diff
@@ -1,46 +1,182 @@
+	}
+	response := wire.Result{ID: id, Error: err}
+	if err := p.sendJSON(response); err != nil {
+		p.Log().WithError(err).Error("Could not send message.")
```
Why do you handle the error here and not return it to the caller? You call it in `readMessage` where you also handle the error of `ReadMessage`.

ggwpez replied:
Well ok, does not make a big difference imho, i changed it.

sebastianst commented on operator/rpc.go:
```diff
@@ -1,46 +1,182 @@
+func (p *Peer) sendJSON(obj interface{}) error {
+	p.connMtx.Lock()
+	defer p.connMtx.Unlock()
+	return p.conn.WriteJSON(obj)
```
So websocket conns are definitely not thread safe?

ggwpez replied:
The gorilla implementation of them is not, the race detector is unhappy without the mutex.

sebastianst commented on operator/rpc.go:
```diff
@@ -1,46 +1,182 @@
+					Proof: proof,
+				}
+			case <-p.sub.quit:
+				break
```
Add logging that the subscription routine returns due to a closed sub.

ggwpez replied:
Added `p.Log().Debug("Subscription routine returns due to closed sub.")`.

sebastianst commented on operator/test/rpcoperator.go:
```diff
@@ -0,0 +1,82 @@
+	//
+	// There are two functions for setting errors that should be returned
+	// when the mocked functions are called.
+	RPCOperator struct {
```
if we rename the `operator.RPCOperator` we should do it similarly here.

ggwpez replied:
We renamed the `RemoteOperator` interface to `WireAPI`, the implementing `RPCOperator` keeps its name, unless you also want to change that.

sebastianst commented on operator/test/rpcoperator.go:
```diff
@@ -0,0 +1,82 @@
+func (r *RPCOperator) Run() {
+	go func() {
+		for {
+			proofs, _ := r.enclave.DepositProofs()
```
Why don't you check the error, log it if not nil and return?

ggwpez replied:
The subscriptions returned by the mocked enclave never return errors, so i omitted a check here (but added a comment :laughing:).

sebastianst commented on operator/test/rpcoperator.go:
```diff
@@ -0,0 +1,82 @@
+	}()
+	go func() {
+		for {
+			proofs, _ := r.enclave.BalanceProofs()
```
Why don't you check the error, log it if not nil and return?

sebastianst commented on operator/test/rpcoperator.go:
```diff
@@ -0,0 +1,82 @@
+// a Transaction to the enclave.
+// Can be read back from Transactions() which bufferes one TX.
+// Returns the error that was set by SetAddTransactionError.
+func (r *RPCOperator) AddTransaction(tx tee.Transaction) error {
```
Rename to `Send`...

sebastianst commented on operator/test/enclave.go:
```diff
@@ -0,0 +1,96 @@
+	}
+}
+
+func (e *Enclave) Shutdown() {
+}
```
remove newline

sebastianst commented on client/rpc_test.go:
```diff
@@ -0,0 +1,122 @@
+	op.Run()
+	rpcServer := operator.NewRPC(op)
+	go func() {
+		if err := rpcServer.Serve("", 8401); err != nil {
```
Put port in `const`

ggwpez replied:
Added `opRPCPort`.

sebastianst commented on client/rpc_test.go:
```diff
@@ -0,0 +1,122 @@
+	}()
+
+	time.Sleep(shortWait)
+	client, err := client.NewRPC("0.0.0.0", 8401)
```
use port const here

sebastianst commented on client/rpc_test.go:
```diff
@@ -0,0 +1,122 @@
+// The construction looks a bit like this:
+// Client <-> RPC <-> MockedRPCOperator <-> OP
+func TestRPC_ClientOp(t *testing.T) {
+	plogrus.Set(logrus.TraceLevel, &logrus.TextFormatter{ForceColors: true})
```
do we really always want trace logging in ci tests? I'd rather set this to `Info` and only use trace during debugging.

ggwpez replied:
Removed it.

sebastianst commented on client/rpc_test.go:
```diff
@@ -0,0 +1,122 @@
+		op.SetSubscribeProofsError(myErr)
+		assert.Error(t, client.Subscribe(ctx, user1))
+		op.SetSubscribeProofsError(nil)
+	})
```
better add newlines after `t.Run` calls

ggwpez committed:
:sparkles: Add tee.Amount wrapper for big.Int.

Amount now wraps big.Int to enable json encoding of big.Int as
decimal string.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:heavy_plus_sign: [cfg] Add gorilla dependency.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

sebastianst commented on tee/tee.go:
```diff
@@ -135,38 +137,26 @@ func (s *Sig) UnmarshalJSON(data []byte) error {
-		"value":   b.Value.Text(10),
-	}
-	return json.Marshal(aux)
+func (s Amount) MarshalJSON() ([]byte, error) {
```
Use `a` as var for all `Amount` methods and pointer.
```suggestion
func (s *Amount) MarshalJSON() ([]byte, error) {
```

ggwpez replied:
PS The marshall needs an object and not a pointer, it silently fails otherwise.

sebastianst commented on operator/test/user.go:
```diff
@@ -165,8 +165,9 @@ func (u *User) BalanceProof(ctx context.Context) {
 
-	if u.bp.Balance.Value.Int64() != u.TargetBalance {
-		u.Errorf("incorrect balance, got %d, expected %d", u.bp.Balance.Value.Int64(), u.TargetBalance)
+	balance := (*big.Int)(u.bp.Balance.Value).Int64()
+	if balance != u.TargetBalance {
```
merge

ggwpez replied:
It is used again one line below, i thought it looks better to only have the `(*big.Int)(u.bp.Balance.Value).Int64()` once.

sebastianst replied:
```suggestion
	if balance := (*big.Int)(u.bp.Balance.Value).Int64(); balance != u.TargetBalance {
```

ggwpez replied:
Ok i did not get it^^

sebastianst commented on operator/rpcoperator.go:
```diff
@@ -0,0 +1,148 @@
+	return *o.subs[addr]
+}
+
+func (o *RPCOperator) PushDepositProof(proof tee.DepositProof) {
```
Do we really need to export the `Push...` methods? I mean, a subscriber outside this package isn't pushing new proofs into the sub, right? Same for balance proofs.

ggwpez replied:
It is used in a Black Box test and must therefore sadly be exported.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,260 @@ package client
+			r.subMtx.RLock()
+			// Don't unlock here, since otherwise a `Subscribe` call could
+			// override the subscription while it receives a proof.
+			defer r.subMtx.RUnlock()
```
Never call `defer`s inside loops. What you're doing here is acquiring indefinitely many read locks and stack indefinitely many unlocks, all executed at once when the `handleConnections` go routine returns.

Also I'm not sure if we even need the `subMtx`. I'll leave a comment about this somewhere else.

ggwpez replied:
🤦

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,260 @@ package client
+// `BalanceProof`.
+func (r *RPC) Subscribe(ctx context.Context, user common.Address) (*Subscription, error) {
+	r.subMtx.Lock()
+	r.subscription = &Subscription{
```
Do we even need the `subMtx`? It looks to me that `RPC.Subscribe` is called once during setup of the client and from then on is used, but never reset, no? Is it even valid to call `Subscribe` twice?

ggwpez replied:
The client assumes it to be valid to subscribe twice, because otherwise that case could not be tested.  
The RPC Server rejects a double subscription obviously.

ggwpez replied:
But i guess since we only call it twice in testing code, i will remove it.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,260 @@ package client
+		id        uint64
+		callbacks map[wire.ID]callback
+
+		subMtx       sync.RWMutex // protects subscription.
```
Not sure if even needed, see other comment.

ggwpez replied:
Removed it since it would only theoretically be needed in testing code.

sebastianst commented on client/rpc.go:
```diff
@@ -4,70 +4,254 @@ package client
+	r.cbMtx.Lock()
+	defer r.cbMtx.Unlock()
+	if _, ok := r.callbacks[id]; ok {
+		r.Log().Error("Callbacks should never be overridden.")
```
"overwritten"

Since you're writing an error to the log here and not return an error, this should be rephrased to a sentence like
```suggestion
		r.Log().WithField("id", id).Error("Callback already registered, skipping.")
```
